### ISSUE TITLE
Support for recursive copy with put?

### ISSUE THREAD

**Issue by zpratt** [mscdex/ssh2-streams#40]
sftp's `put` command supports the `-r` option, which will recursively copy a directory to the destination. After digging through the code, I don't see a way to do that with this lib, but I could be overlooking something. Is it supported? If not, are there plans to support it?

`scp` is not an option for me. I've got to upload files to a server that is secured in such a way that sftp is my only option.

**Comment by mscdex** [comment]
It is not currently supported and have no immediate plans. Usually for cases like this I just stream a tarball (no intermediate tarball file creation) over the ssh connection.

**Comment by zpratt** [comment]
Alright. Looks like I might be able to cook that up with https://github.com/mafintosh/tar-stream. Thanks for the tip? Are there examples of how to do this that I'm overlooking?

**Comment by mscdex** [comment]
I'm not familiar with `tar-stream`, but here is one example using a child process (untested):

```js
var spawn = require('child_process').spawn;

// ...

ssh.exec('cd /path/to/extract/to && tar zxf -', function(err, remote) {
  var local = spawn('tar', ['zcf', '-', '/path/to/dir']);
  remote.stderr.resume();
  remote.stdout.resume();
  remote.on('close', function(code, signal) {
    console.log('Transfer ' + (code === 0 ? 'success' : 'fail'));
  });
  local.stdout.pipe(remote.stdin);
});
```

**Comment by zpratt** [comment]
I do not have shell access to the remote server.

**Comment by zpratt** [comment]
Doesn't look like this is possible when you have sftp-only access. To unpack the streamed tarball on the server, it looks like you need to have shell access.

**Comment by mscdex** [comment]
Ah ok I was not aware that was the case for you. In that case you may look on npm to see if someone has already implemented this or write it from scratch, it should be easy to implement.

**Comment by zpratt** [comment]
I was able to come up with a solution using [node-dir](https://github.com/fshost/node-dir) and the filter and each methods from the async module. My code is crude, but it solves the problem I have right now. Here's a rough outline of what I did:

* figure out the set of sub directories under the target directory
* for each sub directory, first do an `lstat` and make the directory if it does not exist
* after all directories have been created, upload all the files

I'm able to scaffold the directory structure and upload 166 files in ~1 second.
